Dockerfile

```diff
@@ -2,9 +2,9 @@
 # This Dockerfile requires BuildKit to be enabled, by setting the environment variable
 # DOCKER_BUILDKIT=1
 # see https://docs.docker.com/build/buildkit/#getting-started
-ARG BASE_IMAGE="ubuntu:jammy-20230624"
+ARG BASE_IMAGE="ubuntu:jammy-20230804"
 ARG BASE_DIGEST_AMD64="sha256:b060fffe8e1561c9c3e6dea6db487b900100fc26830b9ea2ec966c151ab4c020"
```

Comment:
it fails to update both digests at the same time if also the base images changes 😂 it seems it does the amd64 and arm64 update separately while also including the patch update, whatever runs second fails as the image name already changed and the replacement string is not available anymore
`"replaceString": "ARG BASE_IMAGE=\"ubuntu:jammy-20230624\"\nARG BASE_DIGEST_AMD64=\"sha256:b060fffe8e1561c9c3e6dea6db487b900100fc26830b9ea2ec966c151ab4c020\"`

see full log https://developer.mend.io/github/camunda/zeebe/-/job/3f67a99b-f480-4bd4-983b-d7ffc47e8456

and here the failed replacement for amd64

```
INFO: Cannot find replaceString in current file content. Was it already updated? (branch="renovate/main-ubuntu-0.x")
{
  "baseBranch": "main"
  "packageFile": "Dockerfile"
  "depName": "ubuntu"
  "existingContent": "# syntax=docker/dockerfile:1.4\n# This Dockerfile requires BuildKit to be enabled, by setting the environment variable\n# DOCKER_BUILDKIT=1\n# see https://docs.docker.com/build/buildkit/#getting-started\nARG BASE_IMAGE=\"ubuntu:jammy-20230804\"\nARG BASE_DIGEST_AMD64=\"sha256:b060fffe8e1561c9c3e6dea6db487b900100fc26830b9ea2ec966c151ab4c020\"\nARG BASE_DIGEST_ARM64=\"sha256:cf3cc0848a5d6241b6218bdb51d42be7a9f9bd8c505f3abe1222b9c2ce2451ac\"\nARG JDK_IMAGE=\"eclipse-temurin:17-jdk-jammy\"\nARG JDK_DIGEST_AMD64=\"sha256:b817bf26f878c20996fa0def35fc8b694c2f2785dfe9405c25c07066a2b9595f\"\nARG JDK_DIGEST_ARM64=\"sha256:655a9c9066c9ccc6b9459496203d8e25babb43c552b8f32b4f345d75ef7dc7cc\"\n\n# set to \"build\" to build zeebe from scratch instead of using a distball\nARG DIST=\"distball\"\n\n### AMD64 base image ###\n# BASE_DIGEST_AMD64 is defined at the top of the Dockerfile\n# hadolint ignore=DL3006\nFROM ${BASE_IMAGE}@${BASE_DIGEST_AMD64} as base-amd64\nARG BASE_DIGEST_AMD64\nARG BASE_DIGEST=\"${BASE_DIGEST_AMD64}\"\n\n### ARM64 base image ##\n# BASE_DIGEST_ARM64 is defined at the top of the Dockerfile\n# hadolint ignore=DL3006\nFROM ${BASE_IMAGE}@${BASE_DIGEST_ARM64} as base-arm64\nARG BASE_DIGEST_ARM64\nARG BASE_DIGEST=\"${BASE_DIGEST_ARM64}\"\n\n### Base image ##\n# All package installation, updates, etc., anything with APT should be done here in a single step\n# hadolint ignore=DL3006\nFROM base-${TARGETARCH} as base\nWORKDIR /\n\n# Upgrade all outdated packages and install missing ones (e.g. locales, tini)\n# This breaks reproducibility of builds, but is acceptable to gain access to security patches faster\n# than the base image releases\n# FYI, installing packages via APT also updates the dpkg files, which are few MBs, but removing or\n# caching them could break stuff (like not knowing the package is present) or container scanners\n# hadolint ignore=DL3008\nRUN --mount=type=cache,target=/var/cache/apt,sharing=locked \\\n    --mount=type=cache,target=/var/lib/apt,sharing=locked \\\n    --mount=type=cache,target=/var/log/apt,sharing=locked \\\n    apt-get -qq update && \\\n    apt-get install -yqq --no-install-recommends tini ca-certificates && \\\n    apt-get upgrade -yqq --no-install-recommends\n\n\n### AMD64 JDK image ###\n# JDK_DIGEST_AMD64 is defined at the top of the Dockerfile\n# hadolint ignore=DL3006\nFROM ${JDK_IMAGE}@${JDK_DIGEST_AMD64} as jdk-amd64\n\n### ARM64 JDK image ##\n# JDK_DIGEST_ARM64 is defined at the top of the Dockerfile\n# hadolint ignore=DL3006\nFROM ${JDK_IMAGE}@${JDK_DIGEST_ARM64} as jdk-arm64\n\n### Build custom JRE using the base JDK image\n# hadolint ignore=DL3006\nFROM jdk-${TARGETARCH} as jre-build\n\n# Build a custom JRE which will strip down and compress modules to end up with a smaller Java \\\n# distribution than the official JRE. This will also include useful debugging tools like\n# jcmd, jmod, jps, etc., which take little to no space. Anecdotally, compressing modules add around\n# 10ms to the start up time, which is negligible considering our application takes ~10s to start up.\n# See https://adoptium.net/blog/2021/10/jlink-to-produce-own-runtime/\n# hadolint ignore=DL3018\nRUN jlink \\\n     --add-modules ALL-MODULE-PATH \\\n     --strip-debug \\\n     --no-man-pages \\\n     --no-header-files \\\n     --compress=2 \\\n     --output /jre && \\\n   rm -rf /jre/lib/src.zip\n\n### Java base image\nFROM base AS java\nWORKDIR /\n\n# Inherit from previous build stage\nARG JAVA_HOME=/opt/java/openjdk\n\n# Default to UTF-8 file encoding\nENV LANG='C.UTF-8' LC_ALL='C.UTF-8'\n\n# Setup JAVA_HOME and binaries in the path\nENV JAVA_HOME ${JAVA_HOME}\nENV PATH $JAVA_HOME/bin:$PATH\n\n# Copy JRE from previous build stage\nCOPY --from=jre-build /jre ${JAVA_HOME}\n\n# https://github.com/docker-library/openjdk/issues/212#issuecomment-420979840\n# https://openjdk.java.net/jeps/341\n# TL;DR generate some class data sharing for faster load time\nRUN java -Xshare:dump;\n\n### Build zeebe from scratch ###\nFROM java as build\nWORKDIR /zeebe\nENV MAVEN_OPTS -XX:MaxRAMPercentage=80\nCOPY --link . ./\nRUN --mount=type=cache,target=/root/.m2,rw \\\n    ./mvnw -B -am -pl dist package -T1C -D skipChecks -D skipTests && \\\n    mv dist/target/camunda-zeebe .\n\n### Extract zeebe from distball ###\n# hadolint ignore=DL3006\nFROM base as distball\nWORKDIR /zeebe\nARG DISTBALL=\"dist/target/camunda-zeebe-*.tar.gz\"\nCOPY --link ${DISTBALL} zeebe.tar.gz\n\n# Remove zbctl from the distribution to reduce CVE related maintenance effort w.r.t to containers\nRUN mkdir camunda-zeebe && \\\n    tar xfvz zeebe.tar.gz --strip 1 -C camunda-zeebe && \\\n    find . -type f -name 'zbctl*' -delete\n\n### Image containing the zeebe distribution ###\n# hadolint ignore=DL3006\nFROM ${DIST} as dist\n\n### Application Image ###\n# TARGETARCH is provided by buildkit\n# https://docs.docker.com/engine/reference/builder/#automatic-platform-args-in-the-global-scope\n# hadolint ignore=DL3006\nFROM java as app\n# leave unset to use the default value at the top of the file\nARG BASE_IMAGE\nARG BASE_DIGEST\nARG VERSION=\"\"\nARG DATE=\"\"\nARG REVISION=\"\"\n\n# OCI labels: https://github.com/opencontainers/image-spec/blob/main/annotations.md\nLABEL org.opencontainers.image.base.digest=\"${BASE_DIGEST}\"\nLABEL org.opencontainers.image.base.name=\"docker.io/library/${BASE_IMAGE}\"\nLABEL org.opencontainers.image.created=\"${DATE}\"\nLABEL org.opencontainers.image.authors=\"zeebe@camunda.com\"\nLABEL org.opencontainers.image.url=\"https://zeebe.io\"\nLABEL org.opencontainers.image.documentation=\"https://docs.camunda.io/docs/self-managed/zeebe-deployment/\"\nLABEL org.opencontainers.image.source=\"https://github.com/camunda/zeebe\"\nLABEL org.opencontainers.image.version=\"${VERSION}\"\n# According to https://github.com/opencontainers/image-spec/blob/main/annotations.md#pre-defined-annotation-keys\n# and given we set the base.name and base.digest, we reference the manifest of the base image here\nLABEL org.opencontainers.image.ref.name=\"${BASE_IMAGE}\"\nLABEL org.opencontainers.image.revision=\"${REVISION}\"\nLABEL org.opencontainers.image.vendor=\"Camunda Services GmbH\"\nLABEL org.opencontainers.image.licenses=\"(Apache-2.0 AND LicenseRef-Zeebe-Community-1.1)\"\nLABEL org.opencontainers.image.title=\"Zeebe\"\nLABEL org.opencontainers.image.description=\"Workflow engine for microservice orchestration\"\n\n# OpenShift labels: https://docs.openshift.com/container-platform/4.10/openshift_images/create-images.html#defining-image-metadata\nLABEL io.openshift.tags=\"bpmn,orchestration,workflow\"\nLABEL io.k8s.description=\"Workflow engine for microservice orchestration\"\nLABEL io.openshift.non-scalable=\"false\"\nLABEL io.openshift.min-memory=\"512Mi\"\nLABEL io.openshift.min-cpu=\"1\"\n\nENV ZB_HOME=/usr/local/zeebe \\\n    ZEEBE_BROKER_GATEWAY_NETWORK_HOST=0.0.0.0 \\\n    ZEEBE_STANDALONE_GATEWAY=false \\\n    ZEEBE_RESTORE=false\nENV PATH \"${ZB_HOME}/bin:${PATH}\"\n# Disable RocksDB runtime check for musl, which launches `ldd` as a shell process\n# We know there's no need to check for musl on this image\nENV ROCKSDB_MUSL_LIBC=false\n\nWORKDIR ${ZB_HOME}\nEXPOSE 26500 26501 26502\nVOLUME /tmp\nVOLUME ${ZB_HOME}/data\nVOLUME ${ZB_HOME}/logs\n\nRUN groupadd -g 1000 zeebe && \\\n    adduser -u 1000 zeebe --system --ingroup zeebe && \\\n    chmod g=u /etc/passwd && \\\n    # These directories are to be mounted by users, eagerly creating them and setting ownership\n    # helps to avoid potential permission issues due to default volume ownership.\n    mkdir ${ZB_HOME}/data && \\\n    mkdir ${ZB_HOME}/logs && \\\n    chown -R 1000:0 ${ZB_HOME} && \\\n    chmod -R 0775 ${ZB_HOME}\n\nCOPY --link --chown=1000:0 docker/utils/startup.sh /usr/local/bin/startup.sh\nCOPY --from=dist --chown=1000:0 /zeebe/camunda-zeebe ${ZB_HOME}\n\nUSER zeebe:zeebe\n\nENTRYPOINT [\"tini\", \"--\", \"/usr/local/bin/startup.sh\"]\n"
  "replaceString": "ARG BASE_IMAGE=\"ubuntu:jammy-20230624\"\nARG BASE_DIGEST_AMD64=\"sha256:b060fffe8e1561c9c3e6dea6db487b900100fc26830b9ea2ec966c151ab4c020\""
```

Reply:
@npepinpe seems we hit another limitation of renovate here, alternatives I see are:

1. remove the `BASE_IMAGE` arg and inline it on the amd64/arm64 image, this way renovate can update both without a conflict
2. just use the `jammy` tag again, as we have digest pinning anyway we will always use the latest, we can then also remove the custom versioning rule again

I tend to the 2nd option as it's simpler without any real downside? wdyt?

Reply:
I think the second one is good. Let's just document the reasoning in the file, since it goes against Camunda best practices